adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+		"<img src='" + imageInfoUrl + "' width='" + strconv.Itoa(int(adWidth)) + "' height='" + strconv.Itoa(int(adHeight)) + "'/> " +
+		"</a> " +
+		dspImpTrackings2StrImg.String() +
+		"<script type=\"text/javascript\">" +
+		"var dspClickTrackings = [" + dspClickTrackings + "];" +
+		"function sendGetReq() {" +
+		"sendSomeGetReq(dspClickTrackings)" +
+		"}" +
+		"function sendOneGetReq(url) {" +
+		"var req = new XMLHttpRequest();" +
+		"req.open('GET', url, true);" +
+		"req.send(null);" +
+		"}" +
+		"function sendSomeGetReq(urls) {" +
+		"for (var i = 0; i < urls.length; i++) {" +
+		"sendOneGetReq(urls[i]);" +
+		"}" +
+		"}" +
+		"</script>"
+	return adm, adWidth, adHeight, nil
+}
+
+func getDspImpClickTrackings(content *content) (dspImpTrackings []string, dspClickTrackings string) {
+	if content.Monitor != nil {
+		for _, monitor := range content.Monitor {
+			if monitor.EventType == "imp" && len(monitor.Url) != 0 {
+				dspImpTrackings = monitor.Url
+			}
+			if monitor.EventType == "click" && len(monitor.Url) != 0 {
+				dspClickTrackings = getStrings(monitor.Url)
+			}
+		}
+	}
+	return dspImpTrackings, dspClickTrackings
+}
+
+// Add url for monitor
+func getStrings(eles []string) string {
+	if len(eles) == 0 {
+		return ""
+	}
+	var strs strings.Builder
+	for i := 0; i < len(eles); i++ {
+		strs.WriteString("\"" + eles[i] + "\"")
+		if i < len(eles)-1 {
+			strs.WriteString(",")
+		}
+	}
+	return strs.String()
+}
+
+// get duration, millisecond -> format: 00:00:00.000
+func getDuration(duration int64) string {
+	// duration millisecond
+	if duration == 0 {
+		return "00:00:00.000"
+	}
+	var totalSec = int(duration) / 1000
+	var mmm = int(duration) % 1000
+	var hour = totalSec / 3600
+	var tmp = totalSec % 3600
+	var min = tmp / 60
+	var sec = tmp % 60
+	var result string
+	result += addZero(strconv.Itoa(hour), ":", false)
+	result += addZero(strconv.Itoa(min), ":", false)
+	result += addZero(strconv.Itoa(sec), ".", false)
+	result += addZero(strconv.Itoa(mmm), "", true)
+	return result
+}
+
+func addZero(str string, dot string, isMmm bool) (result string) {
+	if isMmm == false {
+		if len(str) < 2 {
+			result = "0" + str + dot
+		} else {
+			result = str + dot
+		}
+	} else {
+		if len(str) == 1 {
+			result = "00" + str
+		} else if len(str) == 2 {
+			result = "0" + str
+		} else {
+			result = str
+		}
+	}
+	return result
+}
+
+// get field adm for video, vast 3.0
+func (a *adapter) extractAdmVideo(adType int32, content *content, bidType openrtb_ext.BidType, opentrb2Imp openrtb2.Imp) (adm string,
+	adWidth int64, adHeight int64, err error) {
+	if content == nil {
+		return "", 0, 0, fmt.Errorf("extractAdmVideo: content is empty")
+	}
+
+	var clickUrl = ""
+	if content.MetaData.ClickUrl != "" {
+		clickUrl = content.MetaData.ClickUrl
+	} else if content.MetaData.Intent != "" {
+		clickUrl, _ = getDecodeValue(content.MetaData.Intent)
+	}
+
+	var mime = "video/mp4"
+	var resourceUrl = ""
+	var duration = ""
+	if adType == roll {
```

Comment:
Please add test coverage for `adType == roll`.

Reply:
add video.json